include/boost/url/detail/impl/url_template.ipp

```diff
@@ -0,0 +1,963 @@
+                    return rv.error();
+                u.path = *rv;
+            }
+            return u;
+        }
+
+        // authority
+        if( it[0] == '/' &&
+            it[1] == '/')
+        {
+            // "//" always indicates authority
+            it += 2;
+            auto rv = grammar::parse(
+                it, end,
+                authority_template_rule);
+            // authority is allowed to be empty
+            BOOST_ASSERT(rv);
+            u.has_authority = true;
+            u.has_user = rv->has_user;
+            u.user = rv->user;
+            u.has_pass = rv->has_pass;
+            u.pass = rv->pass;
+            u.host = rv->host;
+            u.has_port = rv->has_port;
+            u.port = rv->port;
+        }
+
+        // the authority requires an absolute path
+        // or an empty path
+        if (it == end ||
+            (u.has_authority &&
+             (*it != '/' &&
+              *it != '?' &&
+              *it != '#')))
+        {
+            // path-empty
+            return u;
+        }
+
+        // path-abempty
+        // consume the whole path at once because
+        // we're going to count number of segments
+        // later after the replacements happen
+        static constexpr auto segment_fmt_rule =
+            pct_encoded_fmt_string_rule(path_chars);
+        auto rp = grammar::parse(
+            it, end, segment_fmt_rule);
+        // path-abempty is allowed to be empty
+        BOOST_ASSERT(rp);
+        u.path = *rp;
+
+        // [ "?" query ]
+        {
+            static constexpr auto query_fmt_rule =
+                pct_encoded_fmt_string_rule(query_chars);
+            auto rv = grammar::parse(
+                it, end,
+                grammar::tuple_rule(
+                    grammar::squelch(
+                        grammar::delim_rule('?')),
+                    query_fmt_rule));
+            // query is allowed to be empty but
+            // delim rule is not
+            if (rv)
+            {
+                u.has_query = true;
+                u.query = *rv;
+            }
+        }
+
+        // [ "#" fragment ]
+        {
+            static constexpr auto frag_fmt_rule =
+                pct_encoded_fmt_string_rule(fragment_chars);
+            auto rv = grammar::parse(
+                it, end,
+                grammar::tuple_rule(
+                    grammar::squelch(
+                        grammar::delim_rule('#')),
+                    frag_fmt_rule));
+            // frag is allowed to be empty but
+            // delim rule is not
+            if (rv)
+            {
+                u.has_frag = true;
+                u.frag = *rv;
+            }
+        }
+
+        return u;
+    }
+};
+
+constexpr url_template_rule_t url_template_rule{};
+
+result<url_template>
+parse_url_template(
+    string_view s)
+{
+    return grammar::parse(
+        s, url_template_rule);
+}
+
+} // detail
+} // urls
+} // boost
+
+#endif
```

Comment:
ugh.. another thousand-liner

Reply:
This one is an `.ipp` file though. No intermediary rules for the pattern are included in public headers. 

Anyway, the file contains:

- pattern_rule (172 lines)
- intermediate rules for pattern_rule (334 lines)
- function to apply a pattern to a url_base (305 lines)

The rules are quite long because they are basically implementing more complex versions of each url rule. 

To what level of granularity should I break this?